pkg_resources/__init__.py

```diff
@@ -1582,6 +1703,8 @@ def run_script(self, script_name: str, namespace: dict[str, Any]):
                     **locals()
                 ),
             )
+        if not self.egg_info:
+            raise TypeError("Provider is missing egg_info", self.egg_info)
```

Comment:
This if never runs, right?

The reasoning is that when `self.egg_info` is falsey, then `self.has_metadata(script)` is also falsey, so the first exception always run and the if is never reached...

Reply:
This check could be hit if a subclass overwrote `has_metadata` in a way that doesn't garrantee `self.egg_info` is truthy or that `has_metadata` called by `run_script` will always fail (like `FileMetadata` could hit it if it didn't have the `name == 'PKG-INFO'` check since the base `run_script` forces `name` to start with `scripts/`.

There's just no way to statically communicate that `has_metadata` checks for `self.egg_info` (it could be a type guard if `egg_info` was a parameter of `has_metadata`).

Funnily enough, the `Metadata` test class in `pkg_resources/tests/test_resources.py` shows exactly how this can be hit:
```py
>>> from pkg_resources.tests.test_resources import Metadata
>>> Metadata(["scripts/",""]).run_script("", {})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "setuptools\pkg_resources\__init__.py", line 1708, in run_script
    script_filename = self._fn(self.egg_info, script)
  File "setuptools\pkg_resources\__init__.py", line 1744, in _fn
    return os.path.join(base, *resource_name.split('/'))
  File "C:\Program Files\Python39\lib\ntpath.py", line 78, in join
    path = os.fspath(path)
TypeError: expected str, bytes or os.PathLike object, not NoneType
```


I can alternatively do `if not self.has_metadata(script) or not self.egg_info:`, but I went with a `TypeError` to keep the exact same behaviour just in case someone downstream did override `has_metadata` in a way that triggers this error.

Reply:
Ok, I see, the problem is related to the way the implementation is forcing `os.path.join(self.egg_info, ...)` because in theory if `has_metadata/get_metadata` are implemented by a subclass in a way that does not depend `egg_info`, that would be fine...

I suppose that if the subclass overwrites `_fn` so that `os.path.join` is not called, then it would also be OK.

Reply:
This also brings the question, could `egg_info` be initialized to an empty string instead of None ? All of setuptool's code checks for Truthynes of it, never for an explicit `None`. It's also an issue that will come up again as much code that currently isn't type-checked by mypy (functions are missing annotations) or my pyright branch (error currently disabled to pass as-is initially) makes the assumption that `egg_info` is not None w/o checking.

That would technically count as a breaking change downstream though (if someone did `x.egg_info is not None`). So that can be a conversation for later.

Reply:
I feel that we should not add this exception in this part of the code.

If the major concern is that inheritance can be used so that `self.egg_info` is `None`. Then inheritance also can be used to change `def _fn` to handle the `None` case right? (and all the other methods that rely on `self.egg_info`).

Effectively adding this hard check here is a backwards incompatible change...

---
If inheritance is out of the table, then `self.egg_info` is ensure to not be `None` there.

> This also brings the question, could egg_info be initialized to an empty string instead of None ?

Haven't tested that. Effectively Python functions treat the `""` directory as CWD right? So maybe it would work... but I don't know.

Reply:
Last commit applies your suggestion.

We could consider coercing `None` to `""` in `_fn` since that seems to follow the logic where most code checks for an empty string first, and `run_script` was already letting and empty string pass. But is technically a change in behaviour.

Reply:
Alternatively if this is the only point of contention left for this PR, I can undo the parameter change for `_fn` since it's not public and will get the checker passing. Splitting off the conversation about what to do with a none `egg_info` in `run_script` or `None` base in `_fn`

Reply:
Thank you!

I think we can leave the implementation as per your last commit indefinitely.